File: backend/models/Trip.js
const mongoose = require('mongoose');

const advanceSchema = new mongoose.Schema(
  {
    amount: { type: Number, required: true, min: 0 },
    addedByRole: { type: String, enum: ['owner', 'employee', 'driver'], required: true },
    addedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  },
  { timestamps: true }
);

const tripSchema = new mongoose.Schema(
  {
    pickupDateTime: { type: Date, required: true },
    customerName: { type: String, required: true, trim: true },
    customerMobile: { type: String, required: true, trim: true },
    pickupLocation: { type: String, required: true },
    placesToVisit: [{ type: String, required: true }],
    numberOfDays: { type: Number, required: true, min: 1 },
    driverId: { type: mongoose.Schema.Types.ObjectId, ref: 'Driver', required: true },
    vehicleId: { type: mongoose.Schema.Types.ObjectId, ref: 'Vehicle', required: true },
    startTime: { type: Date },
    endTime: { type: Date },
    startKm: { type: Number, min: 0 },
    endKm: { type: Number, min: 0 },
    tollApplicable: { type: Boolean, default: false },
    permitApplicable: { type: Boolean, default: false },
    parkingApplicable: { type: Boolean, default: false },
    driverBataAssigned: { type: Number, default: 0, min: 0 },
    driverBataAssignedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    driverBataAssignedAt: { type: Date },
    driverBataCredited: { type: Boolean, default: false },
    status: {
      type: String,
      enum: ['scheduled', 'in_progress', 'completed'],
      default: 'scheduled',
    },
    advances: [advanceSchema],
    createdBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  },
  { timestamps: true }
);

module.exports = mongoose.model('Trip', tripSchema);

File: flutter_app/lib/models/trip.dart
class TripModel {
  final String id;
  final String customerName;
  final String customerMobile;
  final String pickupLocation;
  final String status;
  final int numberOfDays;
  final String? driverName;
  final String? vehicleNumber;
  final double driverBataAssigned;

  const TripModel({
    required this.id,
    required this.customerName,
    required this.customerMobile,
    required this.pickupLocation,
    required this.status,
    required this.numberOfDays,
    this.driverName,
    this.vehicleNumber,
    this.driverBataAssigned = 0,
  });

  factory TripModel.fromJson(Map<String, dynamic> json) {
    final driverRaw = json['driverId'];
    final vehicleRaw = json['vehicleId'];

    return TripModel(
      id: json['_id'] as String,
      customerName: (json['customerName'] as String?) ?? '-',
      customerMobile: (json['customerMobile'] as String?) ?? '-',
      pickupLocation: json['pickupLocation'] as String,
      status: json['status'] as String,
      numberOfDays: (json['numberOfDays'] as num).toInt(),
      driverName: driverRaw is Map<String, dynamic> ? (driverRaw['name'] as String?) : null,
      vehicleNumber: vehicleRaw is Map<String, dynamic> ? (vehicleRaw['number'] as String?) : null,
      driverBataAssigned: (json['driverBataAssigned'] as num?)?.toDouble() ?? 0,
    );
  }
}
